### PR TITLE
Fix adding hypervisors

### PR DESCRIPTION
**Pull request by m1kra**
This pull request adds validation of assigned VM hypervisor, such that it can only be a ~~virtual server~~ cloud host or data center asset. This resolves issues #3259 and #3236 . Also see RAS-1181.

### CODE REVIEW

File: src/ralph/virtual/admin.py [hunk 1]
```diff
@@ -52,11 +54,30 @@ class VirtualServerTypeForm(RalphAdmin):
...
+    HYPERVISOR_TYPE_ERR_MSG = 'Hypervisor must be one of \
+                                   DataCenterAsset or VirtualServer.'
+    HYPERVISOR_IS_SELF_ERR_MSG = 'Hypervisor and host must be different.'
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         if 'parent' in self.fields:
             self.fields['parent'].required = True
 
+    def is_valid(self):
```
Comment (mkurek) [inline src/ralph/virtual/admin.py]: isn't it possible to use (create) `validate_parent` method?

File: src/ralph/virtual/admin.py [hunk 2]
```diff
@@ -52,11 +54,30 @@ class VirtualServerTypeForm(RalphAdmin):
...
+        valid = super().is_valid()
+        virtual_server = ContentType.objects.get_for_model(VirtualServer)
+        dc_asset = ContentType.objects.get_for_model(DataCenterAsset)
+        selected_object = self.cleaned_data['parent']
+        if selected_object.content_type not in (virtual_server, dc_asset):
```
Comment (mkurek) [inline src/ralph/virtual/admin.py]: why do we allow to VM be parent of another VM?

File: src/ralph/virtual/admin.py [hunk 3]
```diff
@@ -52,10 +55,28 @@ class VirtualServerTypeForm(RalphAdmin):
...
 class VirtualServerForm(RalphAdminForm):
+    HYPERVISOR_TYPE_ERR_MSG = 'Hypervisor must be one of \
+                                   DataCenterAsset or VirtualServer.'
+    HYPERVISOR_IS_SELF_ERR_MSG = 'Hypervisor and host must be different.'
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         if 'parent' in self.fields:
             self.fields['parent'].required = True
+            self.fields['parent'].validators.extend((
```
Comment (mkurek) [inline src/ralph/virtual/admin.py]: sorry for the confusion - I mean `clean_parent` - if you name a method like that, [it should work out of the box](https://docs.djangoproject.com/en/2.0/ref/forms/validation/), without adding it "manually" to validators

File: src/ralph/virtual/admin.py (hunk 1 above)
Comment (m1kra) [reply src/ralph/virtual/admin.py]: Done, now it's a lot cleaner.

File: src/ralph/virtual/admin.py (hunk 2 above)
Comment (m1kra) [reply src/ralph/virtual/admin.py]: Well, I have changed that to CloudHost. As you have said, `DataCenterAssets` are added as cloud hosts for Ironic.

File: src/ralph/virtual/tests/test_admin.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        self.assertTrue(form.is_valid())
+
+    def test_invalid_hypervisor_type(self):
+        form_data = self._get_basic_form_data()
+        for obj in (
+            BaseObjectFactory(),
+            ServiceEnvironmentFactory(),
+            VirtualServerFactory(),
+            LicenceFactory()
+            ):
```
Comment (mkurek) [inline src/ralph/virtual/tests/test_admin.py]: untab this line

File: src/ralph/virtual/admin.py [hunk 5]
```diff
@@ -52,11 +55,25 @@ class VirtualServerTypeForm(RalphAdmin):
 
 
 class VirtualServerForm(RalphAdminForm):
+    HYPERVISOR_TYPE_ERR_MSG = 'Hypervisor must be one of \
```
Comment (mkurek) [inline src/ralph/virtual/admin.py]: wouldn't it be better to translate it here?

File: src/ralph/virtual/admin.py [hunk 6]
```diff
@@ -52,11 +55,25 @@ class VirtualServerTypeForm(RalphAdmin):
...
         if 'parent' in self.fields:
             self.fields['parent'].required = True
 
+    def clean_parent(self):
+        value = self.cleaned_data.get('parent')
+        self._validate_parent_type(value)
+        return value
+
+    def _validate_parent_type(self, value):
+        dc_asset = ContentType.objects.get_for_model(DataCenterAsset)
```
Comment (mkurek) [inline src/ralph/virtual/admin.py]: you might use `ContentType.get_for_models`

**Comment by m1kra** [comment]
I have been trying to solve it differently - filter out only valid hosts from ui. I didn't want to modify base object, adding `limit_choices_to`, so I tried doing it in the form. The field `parent` is `raw_id_field` in admin, and simple modification of queryset didn't work. I have tried adding custom `ForeignKeyRawIdWidget`, which is used by `raw_id_fields`, but it doesn't render correctly. I think I could try to solve that, but I would need some time to figure out how to do it.

File: src/ralph/virtual/admin.py (hunk 3 above)
Comment (m1kra) [reply src/ralph/virtual/admin.py]: Done.

File: src/ralph/virtual/tests/test_admin.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        return True
+
+
+class MockRequest:
+    pass
+
+
+class TestVirtualServerForm(RalphTestCase):
+    
+    def setUp(self):
```
Comment (mkurek) [inline src/ralph/virtual/tests/test_admin.py]: check out these places for inspiration how to make requests to admin in tests:
https://github.com/allegro/ralph/blob/ng/src/ralph/data_center/tests/test_admin.py#L29-L34
https://github.com/allegro/ralph/blob/ng/src/ralph/data_center/tests/test_admin.py#L69-L77

File: src/ralph/virtual/admin.py (hunk 6 above)
Comment (m1kra) [reply src/ralph/virtual/admin.py]: Done.

File: src/ralph/virtual/tests/test_admin.py (hunk 4 above)
Comment (m1kra) [reply src/ralph/virtual/tests/test_admin.py]: Done.

File: src/ralph/virtual/tests/test_admin.py (hunk 7 above)
Comment (m1kra) [reply src/ralph/virtual/tests/test_admin.py]: Done.

File: src/ralph/virtual/admin.py (hunk 5 above)
Comment (m1kra) [reply src/ralph/virtual/admin.py]: Done.

**Review by mkurek — APPROVED** [review]

**Pull request merged**